README.md
# DWX ZeroMQ Connector  { Python 3 to MetaTrader 4 }

# Latest version: 2.0.1 [(here)](https://github.com/darwinex/dwx-zeromq-connector/tree/master/v2.0.1)
## (v2.0.2 is currently in beta-testing, please do not upgrade to v2.0.2)

## Need help? Join the [Darwinex Collective Slack](https://join.slack.com/t/darwinex-collective/shared_invite/enQtNjg4MjA0ODUzODkyLWFiZWZlMDZjNGVmOGE2ZDBiZGI4ZWUxNjM5YTU0MjZkMTQ2NGZjNGIyN2QxZDY4NjUyZmVlNmU3N2E2NGE1Mjk) for code updates, Q&A and more.

## Table of Contents
* [Introduction](#introduction)
* [Installation](#installation)
* [Configuration](#configuration)
* [Example Usage](#example-usage)
* [Video Tutorials](#video-tutorials)
* [Complete list of available functions](#available-functions)
* [License](#license) 

## Introduction
In this project, we present a technique employing ZeroMQ (an Open Source, Asynchronous Messaging Library and Concurrency Framework) for building a basic – but easily extensible – high performance bridge between external (non-MQL) programming languages and MetaTrader 4.

### IMPORTANT NOTES - PLEASE READ:

1. Please note that we cannot provide support for Python or MQL as programming languages themselves. Therefore, if you are new to Python and MQL, incorporating the project into your specific algorithmic trading environment will require some additional work on your part (i.e. enough Python experience to integrate the Bridge into your environment -> it is assumed that users of the Bridge are self-sufficient in Python).

1. Any code provided and/or referenced in this repository is NOT meant to be used "as-is". Users must treat all code as educational content that requires modification / incorporation into existing works, as per their individual requirements.

1. We have drafted as detailed a set of steps as possible in our project README, but cannot cover all the dependencies as they are independent projects on their own that programmers need to account for / follow / keep up to speed with when considering using the DWX ZeroMQ Connector.

1. This project and all accompanying source code should be run standalone (i.g. via a Python or IPython console, or batch process).

1. Please DO NOT run this code in Jupyter or IPython Notebooks.

1. The project's dependencies require **MS VC++ Libraries**. Without these installed, you are likely to run into **"Resource Timeout" errors.** The DLLs in the dependency projects (mql-zmq, libzmq, libsodium) require that you have the **latest Visual C++ runtime (2015) libraries already installed.**

1. This project has not been tested on emulated environments (e.g. WINE, VMWare, etc).

1. This project is intended for use solely in Windows 10 environments, at the present time.
	
**Reasons for writing this post:**

* Lack of comprehensive, publicly available literature about this topic on the web.
* Traders have traditionally relied on Winsock/WinAPI based solutions that often require revision with both Microsoft™ and MetaQuotes™ updates.
* Alternatives to ZeroMQ include named pipes, and approaches where filesystem-dependent functionality forms the bridge between MetaTrader and external languages.

**We lay the foundation for a distributed trading system that will:**

* Consist of one or more trading strategies developed outside MetaTrader 4 (non-MQL),
* Use MetaTrader 4 for acquiring market data, trade execution and management,
* Support multiple non-MQL strategies interfacing with MetaTrader 4 simultaneously,
* Consider each trading strategy as an independent “Client”,
* Consider MetaTrader 4 as the “Server”, and medium to market,
* Permit both Server and Clients to communicate with each other on-demand.

**Infographic: ZeroMQ-Enabled Distributed Trading Infrastructure (with MetaTrader 4)**
![DWX ZMQ Infographic 1](resources/images/dwx-zeromq-infographic-1.png)

**Why ZeroMQ?**

* Enables programmers to connect any code to any other code, in a number of ways.
* Eliminates a MetaTrader user’s dependency on just MetaTrader-supported technology (features, indicators, language constructs, libraries, etc.)
* Traders can develop indicators and strategies in C/C#/C++, Python, R and Java (to name a few), and deploy to market via MetaTrader 4.
* Leverage machine learning toolkits in Python and R for complex data analysis and strategy development, while interfacing with MetaTrader 4 for trade execution and management.
* ZeroMQ can be used as a high-performance transport layer in sophisticated, distributed trading systems otherwise difficult to implement in MQL.
* Different strategy components can be built in different languages if required, and seamlessly talk to each other over TCP, in-process, inter-process or multicast protocols.
* Multiple communication patterns and disconnected operation.

## Installation

This project requires the following:

* **Python**: (minimum v3.6)
* **libzmq**: (minimum v4.2.5)
* **pyzmq**: (minimum v17.1.2)
* **libsodium** (https://github.com/jedisct1/libsodium)
* **mql4-lib** (https://github.com/dingmaotu/mql4-lib)
* **mql-zmq** (https://github.com/dingmaotu/mql-zmq)

For your convenience, files from the last three items above have been included in this repository with appropriate copyrights referenced within. 

_This project incorporates functionality authored by Ding Li (GitHub: https://github.com/dingmaotu), who has kindly licensed his work under the Apache 2.0 license._

_We acknowledge copyright as per the terms of the license, the following repositories serving as mandatory dependencies for this project:_

1. _https://github.com/dingmaotu/mql-zmq_

1. _https://github.com/dingmaotu/mql4-lib_

_Thank you Ding for your amazing open source contribution to this space!_

_Sincerely,_<br>
_The Darwinex Labs Team_<br>
_www.darwinex.com_


### Steps:

1. Download and unzip **mql-zmq-master.zip** (by GitHub author @dingmaotu)
1. Copy the contents of **mql-zmq-master/Include/Mql** and **mql-zmq-master/Include/Zmq** into your MetaTrader installation's **MQL4/Include** directory as-is. Your **MQL4/Include** directory should now have two additional folders "Mql" and "Zmq".
1. Copy **libsodium.dll** and **libzmq.dll** from **mql-zmq-master/Library/MT4** to your MetaTrader installation's **MQL4/Libraries** directory.
1. Download **DWX_ZeroMQ_Server_vX.Y.Z_RCx.mq4** and place it inside your MetaTrader installation's **MQL4/Experts** directory.
1. Finally, download **vX.Y.Z / python / api / DWX_ZeroMQ_Connector_vX_Y_Z_RCx.py**.

**Note:** vX_Y_Z_RCx refers to version and release candidate, e.g. v2.0.1_RC8.

## Configuration

1. After completing the steps above, terminate and restart MetaTrader 4.
1. Open any new chart, e.g. EUR/USD M, then drag and drop **DWX_ZeroMQ_Server_vX.Y.Z_RCx**.
1. Switch to the EA's Inputs tab and customize values as necessary:

    ![EA Inputs](resources/images/expert-inputs.png)
1. Note: Setting **Publish_MarketData** to **True** will cause MetaTrader 4 to begin publishing BID/ASK tick data in real-time for all symbols specified in the array **Publish_Symbols** contained in the .mq4 script. 
1. Simply modify the Publish_Symbols[] aarray's contents to add/remove required symbols as necessary and re-compile. 
1. The default list of symbols is:

	```
	string Publish_Symbols[7] = {
	   "EURUSD","GBPUSD","USDJPY","USDCAD","AUDUSD","NZDUSD","USDCHF"
	};
	```
	![MetaTrader Publishing Tick Data 1](resources/images/ZeroMQ_Server_Publishing_Symbol_Data.gif)
	
	![MetaTrader Publishing Tick Data 2](resources/images/InAction_ZeroMQ_Server_Publishing_Symbol_Data.gif)

## Example Usage

### Initialize Connector:
```
_zmq = DWX_ZeroMQ_Connector()
```

### Construct Trade to send via ZeroMQ to MetaTrader:
```
_my_trade = _zmq.gen_default_order_dict()

Output: 
{'_action': 'OPEN',
 '_type': 0,
 '_symbol': 'EURUSD',
 '_price': 0.0,
 '_SL': 500,
 '_TP': 500,
 '_comment': 'DWX_Python_to_MT',
 '_lots': 0.01,
 '_magic': 123456,
 '_ticket': 0}

_my_trade['_lots'] = 0.05

_my_trade['_SL'] = 250

_my_trade['_TP'] = 750

_my_trade['_comment'] = 'nerds_rox0r'
```

### Send trade to MetaTrader:
```
_zmq.push_new_order(_order=_my_trade)

# MetaTrader response (JSON):
{'_action': 'EXECUTION', 
'_magic': 123456, 
'_ticket': 85051741, 
'_open_price': 1.14414, 
'_sl': 250, 
'_tp': 750}
```

### Get all open trades from MetaTrader:
```
_zmq.push_req_open_orders()

# MetaTrader response (JSON):

{'_action': 'OPEN_TRADES', 
'_trades': {
    85051741: {'_magic': 123456, 
                '_symbol': 'EURUSD', 
                '_lots': 0.05, 
                '_type': 0, 
                '_open_price': 1.14414, 
                '_pnl': -0.45}
    }
}
```

### Partially close 0.01 lots:
```
_zmq.push_close_order_partial(85051741, 0.01)

# MetaTrader response (JSON):
{'_action': 'CLOSE', 
'_ticket': 85051741, 
'_response': 'CLOSE_PARTIAL', 
'_close_price': 1.14401, 
'_close_lots': 0.01}

# Partially closing a trade renews the ticket ID, retrieve it again.

_zmq.push_req_open_orders()

# MetaTrader response (JSON):
{'_action': 'OPEN_TRADES', 
'_trades': {
    85051856: {'_magic': 123456, 
                '_symbol': 'EURUSD', 
                '_lots': 0.04, 
                '_type': 0, 
                '_open_price': 1.14414, 
                '_pnl': -0.36}
    }
}
```

### Close a trade by ticket:
```
_zmq.push_close_order(85051856)

# MetaTrader response (JSON):
{'_action': 'CLOSE', 
'_ticket': 85051856, 
'_close_price': 1.14378, 
'_close_lots': 0.04, 
'_response': 'CLOSE_MARKET', 
'_response_value': 'SUCCESS'}
```

### Close all trades by Magic Number:
```
# Before running the following example, 5 trades were executed using the same values as in "_my_trade" above, the magic number being 123456.

# Check currently open trades.

_zmq.push_req_open_orders()

# MetaTrader response (JSON):
{'_action': 'OPEN_TRADES', 
    '_trades': {
        85052022: {'_magic': 123456, '_symbol': 'EURUSD', '_lots': 0.05, '_type': 0, '_open_price': 1.14353, '_pnl': 1.15}, 
        85052026: {'_magic': 123456, '_symbol': 'EURUSD', '_lots': 0.05, '_type': 0, '_open_price': 1.14354, '_pnl': 1.1}, 
        85052025: {'_magic': 123456, '_symbol': 'EURUSD', '_lots': 0.05, '_type': 0, '_open_price': 1.14354, '_pnl': 1.1}, 
        85052024: {'_magic': 123456, '_symbol': 'EURUSD', '_lots': 0.05, '_type': 0, '_open_price': 1.14354, '_pnl': 1.1}, 
        85052023: {'_magic': 123456, '_symbol': 'EURUSD', '_lots': 0.05, '_type': 0, '_open_price': 1.14356, '_pnl': 1}
    }
}

# Close all trades with magic number 123456
_zmq.push_close_magic(123456)

# MetaTrader response (JSON):
{'_action': 'CLOSE_ALL_MAGIC', '_magic': 123456, 
'_responses': {
    85052026: {'_symbol': 'EURUSD', '_close_price': 1.14375, '_close_lots': 0.05, '_response': 'CLOSE_MARKET'}, 
    85052025: {'_symbol': 'EURUSD', '_close_price': 1.14375, '_close_lots': 0.05, '_response': 'CLOSE_MARKET'}, 
    85052024: {'_symbol': 'EURUSD', '_close_price': 1.14375, '_close_lots': 0.05, '_response': 'CLOSE_MARKET'}, 
    85052023: {'_symbol': 'EURUSD', '_close_price': 1.14375, '_close_lots': 0.05, '_response': 'CLOSE_MARKET'}, 
    85052022: {'_symbol': 'EURUSD', '_close_price': 1.14375, '_close_lots': 0.05, '_response': 'CLOSE_MARKET'}}, 
'_response_value': 'SUCCESS'}
```

### Close ALL open trades:
```

# Open 5 trades

_zmq.push_new_order()
{'_action': 'EXECUTION', '_magic': 123456, '_ticket': 85090920, '_open_price': 1.15379, '_sl': 500, '_tp': 500}

_zmq.push_new_order()

CHANGELOG.md
# DWX ZeroMQ Connector  { Python 3 to MetaTrader 4 }

# Changes in v2.0.2 compared to v2.0.1:

## Table of Contents
* [Introduction](#introduction)
* [Price feeds](#price-feed)
* [Rates History](#rates-history)
* [Rates feed](#rates-feed)
* [New Data Notifications in Real Time](#new-data-notifications-in-real-time)
* [List of changes](#list-of-changes)
* [Examples](#examples)
---

## Introduction
With this version update, kindly contributed by the amazing [Raul](https://github.com/raulMrello), several modifications have been made to both the Python and MQL4 scripts, to improve the overall capabilities provided by this project. All modifications have been deployed in subfolder **v2.0.2**.

## Price feed

In  version v2.0.1, clients could subscribe to price feeds (```bid,ask``` prices) for different symbols. Those symbols were required to be previouly configured in the Expert Advisor.

However, if clients (in runtime) did a new subscription to other symbols (not configured previously in the Expert Advisor), their price feed could not be updated, because those symbols were not configured yet in the EA.

The modification proposed in v2.0.2 by the pull-request author Raul, required a modification be made to both the Expert Advisor code and ZMQ-Connector code, whereby symbols could be configured in the Expert Advisor (runing as a Server in MT4) and hence, enabling clients to change subscriptions to their price feeds during at any time.

---
## Rates History

In version 2.0.1, clients could request historical prices for specific instruments (symbol & timeframe), and receive ```CLOSE``` prices for a period in time.

However, there wasn't a service to get ```RATE``` prices for those instruments, where a ```RATE``` as a data stream formed by these values: ```TIME,OPEN,HIGH,LOW,CLOSE,TICKVOL,SPREAD,REALVOL```.

This has been implemented in v2.0.2, and includes a new service providing this functionality.

---
## Rates feed

In version 2.0.1, there was no way to get a rates feed for specific instruments. 

The modification proposed in v2.0.2, allowed clients to modify instruments configured in the Expert Advisor and hence get subscribed to their rates feed during runtime.

---
## New Data Notifications in Real Time

In version 2.0.1, data received from the Server came through PULL or SUB ports.

Data feeds came through SUB ports and were registered in the client's dictionary ```self._Market_Data_DB```. So, clients could check this dictionary periodically for new received data.

In v2.0.2 implements a change in the ZMQ-Connector code. Now a data processor is registered in the Connector, acting as an event handler. When new data is received through the PULL port, a callback is invoked ```dataHandler.onPullData(pull_port_data)```.

Similarly, when data is received through the SUB port,  a callback ```dataHandler.onSubData(sub_port_data)``` is invoked.

With the addition of these callbacks, clients may now execute in a more efficient way, via this event-driven framework.

---
## List of changes

This section describes changes to source code and newly added functionalities.

### **_TRACK_PRICES command_**

This request is intended to change symbols configured in the Expert Advisor at runtime. The request is formed using a command and a list of symbols that must be managed by the Expert Advisor:

```cpp
TRACK_PRICES;SYMBOL_1;SYMBOL_2;SYMBOL_3;.....;SYMBOL_N
```

_Example 1: to start receiving bid-ask prices for EURUSD, GDAXI and EURGBP, clients can send this PUSH request to the Server:_

```cpp
TRACK_PRICES;EURUSD;GDAXI;EURGBP
```

_Example 2: At any given time, clients can cancel e.g. the EURUSD price feed, updating that list with this new PUSH request:_

```cpp
TRACK_PRICES;GDAXI;EURGBP
```

_Example 3: Also, clients can clears the symbol list managed by the EA, sending a request without symbols:_

```cpp
TRACK_PRICES
```

To enable this new functionality, a new method was added:

```cpp
def push_req_prices_track(self,_symbols=['EURUSD'])
```

The MQL4 Server EA was also modified to accept this new command request, to provide an associated response and update the list of symbols managed by itself. The response to this request <```TRACK_PRICES;GDAXI;EURGBP```> is as follows:

```cpp
{
  '_action': 'TRACK_PRICES',
  '_data': {
    '_symbol_count': 2
  }
}
```

### **_TRACK_RATES command_**

This request is intended to change instruments configured in the Expert Advisor at runtime, so that clients could get subscribed to rate price feeds. 

It is identified by the command: ```TRACK_RATES``` followed by a list of symbols and timeframes:

```cpp
TRACK_RATES;SYMBOL_1;TIMEFRAME_1;SYMBOL_2;TIMEFRAME_2;.....;SYMBOL_N;TIMEFRAME_N
```

_Example 1: to start receiving Rates(time-open-high-low-close-tickvol-spread-realvol) from EURUSD at timeframe M1, and from GDAXI at timeframe H4, clients can send this PUSH request to the Server:_

```cpp
TRACK_RATES;EURUSD;1;GDAXI;240
```

_Note: H4 = 4hours = 4*60mins = 240mins (timeframe = 240)_

_Example 2: instrument list can be changed at any given time. It can also be cleared by a command without symbols-timeframes:_

```cpp
TRACK_RATES
```

To enable this new functionality, a new method was added:

```cpp
def push_req_rates_track(self,_instruments=[('EURUSD_M1','EURUSD',1)]):
```  

_Note: Instruments are tuples of 3 parameters (name, symbol, timeframe). In this case, name is formatted as: ```SYMBOL_TIMEFRAME```, where ```TIMEFRAME``` holds values from one of the standard MQL4 timeframes: ```M1,M5,M15,M30,H1,H4,D1,W1,MN1```_

The Expert Advisor was modified to process this new request properly. The response to this request <```TRACK_RATES;GDAXI;1```> is as follows:

```cpp
{
  '_action': 'TRACK_RATES',

DWX_ZeroMQ_Connector.py
# -*- coding: utf-8 -*-

"""
    @author: Darwinex Labs (www.darwinex.com)
    
    Copyright (c) 2017-2019, Darwinex. All rights reserved.
    
    Licensed under the BSD 3-Clause License, you may not use this file except 
    in compliance with the License. 
    
    You may obtain a copy of the License at:    
    https://opensource.org/licenses/BSD-3-Clause
"""

import zmq
from time import sleep
from pandas import DataFrame, Timestamp
from threading import Thread
import copy


class ZMQConnector:
    """
    Setup ZeroMQ -> MetaTrader Connector
    """

    ##########################################################################

    def __init__(self,
                 client_id='DLabs_Python',
                 host='localhost',
                 protocol='tcp',
                 push_port=32768,
                 pull_port=32769,
                 sub_port=32770,
                 delimiter=';',
                 pulldata_handlers=None,
                 subdata_handlers=None,
                 verbose=False):

        """
        Parameters
        ----------
        client_id : str
            Unique ID for this client
        host : str
            Host to connect to
        protocol : {'tcp', 'udp'}
            Connection protocol
        push_port : int
            Port for Sending commands
        pull_port : int
            Port for Receiving commands
        sub_port : int
            Port for Subscribing for prices
        delimiter : str
            command strings delimiter
        pulldata_handlers : list of class
            Handlers to process data received through PULL port
        subdata_handlers : list of class
            Handlers to process data received through SUB port
        verbose : bool
        """

        # Strategy Status (if this is False, ZeroMQ will not listen for data)
        self._ACTIVE = True

        # Client ID
        self._CLIENT_ID = client_id

        # ZeroMQ Host
        self._HOST = host

        # Connection Protocol
        self._PROTOCOL = protocol

        # ZeroMQ Context
        self.zmq_context = zmq.Context()

        # TCP Connection URL Template
        self._URL = self._PROTOCOL + "://" + self._HOST + ":"

        # Ports for PUSH, PULL and SUB sockets respectively
        self._PUSH_PORT = push_port
        self._PULL_PORT = pull_port
        self._SUB_PORT = sub_port

        # Handlers for received data (pull and sub ports)
        self.pulldata_handlers = pulldata_handlers or []
        self.subdata_handlers = subdata_handlers or []

        # Create Sockets
        self.push_socket = self.zmq_context.socket(zmq.PUSH)
        self.push_socket.setsockopt(zmq.SNDHWM, 1)

        self.pull_socket = self.zmq_context.socket(zmq.PULL)
        self.pull_socket.setsockopt(zmq.RCVHWM, 1)

        self.sub_socket = self.zmq_context.socket(zmq.SUB)

        # Bind PUSH Socket to send commands to MetaTrader
        self.push_socket.connect(self._URL + str(self._PUSH_PORT))
        print("[INIT] Ready to send commands to METATRADER (PUSH): " + str(self._PUSH_PORT))

        # Connect PULL Socket to receive command responses from MetaTrader
        self.pull_socket.connect(self._URL + str(self._PULL_PORT))
        print("[INIT] Listening for responses from METATRADER (PULL): " + str(self._PULL_PORT))

        # Connect SUB Socket to receive market data from MetaTrader
        self.sub_socket.connect(self._URL + str(self._SUB_PORT))
        print("[INIT] Listening for market data from METATRADER (SUB): " + str(self._SUB_PORT))

        # Initialize POLL set and register PULL and SUB sockets
        self.poller = zmq.Poller()
        self.poller.register(self.pull_socket, zmq.POLLIN)
        self.poller.register(self.sub_socket, zmq.POLLIN)

        # Start listening for responses to commands and new market data
        self._DELIMITER = delimiter

        # BID/ASK Market Data Subscription Threads ({SYMBOL: Thread})
        self._DATA_THREAD = None

        # Begin polling for PULL / SUB data
        self._DATA_THREAD = Thread(target=self._receive_data, args=self._DELIMITER)
        self._DATA_THREAD.start()

        # Market Data Dictionary by Symbol (holds tick data) or Instrument (holds OHLC data)
        self.Data_DB = {}  # {SYMBOL: {TIMESTAMP: (BID, ASK)}}
        # {SYMBOL: {TIMESTAMP: (TIME, OPEN, HIGH, LOW, CLOSE, TICKVOL, SPREAD, VOLUME)}}

        # Temporary Order STRUCT for convenience wrappers later.
        self.temp_order_dict = self.gen_default_order_dict()

        # Thread returns the most recently received DATA block here
        self.thread_data_output = None

        # Verbosity
        self._VERBOSE = verbose

    ##########################################################################
    # internals

    @staticmethod
    def _remote_send(socket, data):
        """
        Function to send commands to MetaTrader (PUSH)
        """

        try:
            socket.send_string(data, zmq.DONTWAIT)
        except zmq.error.Again:
            print("Resource timeout.. please try again.")
            sleep(0.000000001)

    @staticmethod
    def _remote_recv(socket):
        """
        Function to retrieve data from MetaTrader (PULL or SUB)
        """

        try:
            msg = socket.recv_string(zmq.DONTWAIT)
            return msg
        except zmq.error.Again:
            print("Resource timeout.. please try again.")
            sleep(0.000001)

        return None

    def _get_response_(self):
        return self.thread_data_output

    def _set_response_(self, resp=None):
        self.thread_data_output = resp

    def _valid_response_(self, input_='zmq'):
        # Valid data types
        _types = (dict, DataFrame)

        # If input_ = 'zmq', assume self._zmq.thread_data_output
        if isinstance(input_, str) and input_ == 'zmq':
            return isinstance(self._get_response_(), _types)
        else:
            return isinstance(input_, _types)

    def _set_status(self, new_status=False):
        self._ACTIVE = new_status
        print('**')
        print('[KERNEL] Setting Status to {} - Deactivating Threads.. please wait a bit.'.format(new_status))
        print('**')

    def _push_command(self, _action='OPEN', _type=0,
                      _symbol='EURUSD', _price=0.0,
                      _sl=50, _tp=50, _comment="Python-to-MT",
                      _lots=0.01, _magic=123456, _ticket=0):
        """ Function to construct messages for sending Trade commands to MetaTrader
        format:
            compArray[0] = TRADE
            compArray[1] = ACTION (e.g. OPEN, MODIFY, CLOSE)
            compArray[2] = TYPE (e.g. OP_BUY, OP_SELL, etc - only used when ACTION=OPEN)
            // ORDER TYPES:
            // https://docs.mql4.com/constants/tradingconstants/orderproperties
            // OP_BUY = 0
            // OP_SELL = 1
            // OP_BUYLIMIT = 2
            // OP_SELLLIMIT = 3
            // OP_BUYSTOP = 4
            // OP_SELLSTOP = 5
            compArray[3] = Symbol (e.g. EURUSD, etc.)
            compArray[4] = Open/Close Price (ignored if ACTION = MODIFY)
            compArray[5] = SL
            compArray[6] = TP
            compArray[7] = Trade Comment
            compArray[8] = Lots
            compArray[9] = Magic Number
            compArray[10] = Ticket Number (MODIFY/CLOSE)
        """

        _msg = "{};{};{};{};{};{};{};{};{};{};{}".format('TRADE', _action, _type,
                                                         _symbol, _price,
                                                         _sl, _tp, _comment,
                                                         _lots, _magic,
                                                         _ticket)

        # Send via PUSH Socket
        self._remote_send(self.push_socket, _msg)

    def _receive_data(self, string_delimiter=';'):
        """Function to check Poller for new reponses (PULL) and market data (SUB)"""

        while self._ACTIVE:
            sockets = dict(self.poller.poll())

            # Process response to commands sent to MetaTrader
            if self.pull_socket in sockets and sockets[self.pull_socket] == zmq.POLLIN:
                try:
                    msg = self.pull_socket.recv_string(zmq.DONTWAIT)

                    # If data is returned, store as pandas Series
                    if msg != '' and msg is not None:
                        try:
                            _data = eval(msg)
                            self.thread_data_output = _data
                            if self._VERBOSE:
                                print(_data)  # default logic

                            # invokes data handlers on pull port
                            for hnd in self.pulldata_handlers:
                                hnd.onPullData(_data)
                        except Exception as ex:
                            _exstr = "Exception Type {0}. Args:\n{1!r}"
                            _msg = _exstr.format(type(ex).__name__, ex.args)
                            print(_msg)
                except zmq.error.Again:
                    pass  # resource temporarily unavailable, nothing to print
                except ValueError:
                    pass  # No data returned, passing iteration.
                except UnboundLocalError:
                    pass  # _symbol may sometimes get referenced before being assigned.

            # Receive new market data from MetaTrader
            if self.sub_socket in sockets and sockets[self.sub_socket] == zmq.POLLIN:
                try:
                    msg = self.sub_socket.recv_string(zmq.DONTWAIT)
                    if msg != "":
                        _timestamp = str(Timestamp.now('UTC'))[:-6]
                        _symbol, _data = msg.split(" ")
                        if len(_data.split(string_delimiter)) == 2:
                            _bid, _ask = _data.split(string_delimiter)
                            if self._VERBOSE:
                                print("\n[" + _symbol + "] " + _timestamp + " (" + _bid + "/" + _ask + ") BID/ASK")

                                # Update Market Data DB
                            if _symbol not in self.Data_DB.keys():
                                self.Data_DB[_symbol] = {}
                            self.Data_DB[_symbol][_timestamp] = (float(_bid), float(_ask))
                        elif len(_data.split(string_delimiter)) == 8:
                            _time, _open, _high, _low, _close, _tick_vol, _spread, _real_vol = _data.split(
                                string_delimiter)
                            if self._VERBOSE:
                                print_string = "\n[" + _symbol + "] " + _timestamp + " (" + _time + "/"
                                print_string += _open + "/" + _high + "/" + _low + "/" + _close + "/" + _tick_vol + "/"
                                print_string += _spread + "/" + _real_vol
                                print_string += ") TIME/OPEN/HIGH/LOW/CLOSE/TICKVOL/SPREAD/VOLUME"
                                print(print_string)

                            # Update Market Rate DB
                            if _symbol not in self.Data_DB.keys():
                                self.Data_DB[_symbol] = {}
                            self.Data_DB[_symbol][_timestamp] = (
                                int(_time), float(_open), float(_high), float(_low), float(_close), int(_tick_vol),
                                int(_spread), int(_real_vol))

                        # invokes data handlers on sub port
                        for hnd in self.subdata_handlers:
                            hnd.onSubData(msg)
                except zmq.error.Again:
                    pass  # resource temporarily unavailable, nothing to print
                except ValueError:
                    pass  # No data returned, passing iteration.
                except UnboundLocalError:
                    pass  # _symbol may sometimes get referenced before being assigned.

    def wait_for_pull(self, timeout=1):
        wait_s = 0.01
        rec = None
        for _ in range(int(timeout / wait_s)):
            if self.thread_data_output is None:
                sleep(wait_s)
                continue
            rec = copy.deepcopy(self.thread_data_output)
            break
        self.thread_data_output = None
        return rec

    ##########################################################################
    # Default formats

    @staticmethod
    def gen_default_order_dict():
        return ({'_action': 'OPEN',
                 '_type': 0,
                 '_symbol': 'EURUSD',
                 '_price': 0.0,
                 '_SL': 500,  # SL/TP in POINTS, not pips.
                 '_TP': 500,
                 '_comment': 'DWX_Python_to_MT',
                 '_lots': 0.01,
                 '_magic': 123456,
                 '_ticket': 0})

    @staticmethod
    def gen_default_data_dict():
        return ({'_action': 'DATA',
                 '_symbol': 'EURUSD',
                 '_timeframe': 1440,  # M1 = 1, M5 = 5, and so on..
                 '_start': '2018.12.21 17:00:00',  # timestamp in MT4 recognized format
                 '_end': '2018.12.21 17:05:00'})

    @staticmethod
    def gen_default_hist_dict():
        return ({'_action': 'HIST',
                 '_symbol': 'EURUSD',
                 '_timeframe': 1,  # M1 = 1, M5 = 5, and so on..
                 '_start': '2018.12.21 17:00:00',  # timestamp in MT4 recognized format
                 '_end': '2018.12.21 17:05:00'})

    ##########################################################################
    # Convenience functions to permit easy trading via underlying functions.

    def push_new_order(self, order=None):
        """open a new order"""

        if order is None:
            order = self.gen_default_order_dict()

        # Execute
        self._push_command(**order)

    def push_modify_order(self, ticket, sl, tp):  # in points
        """ modify an order"""

        try:
            self.temp_order_dict['_action'] = 'MODIFY'
            self.temp_order_dict['_SL'] = sl
            self.temp_order_dict['_TP'] = tp
            self.temp_order_dict['_ticket'] = ticket

            # Execute
            self._push_command(**self.temp_order_dict)

        except KeyError:
            print("[ERROR] Order Ticket {} not found!".format(ticket))

    def push_close_order(self, ticket):
        """ close an order"""

        try:
            self.temp_order_dict['_action'] = 'CLOSE'
            self.temp_order_dict['_ticket'] = ticket

            # Execute
            self._push_command(**self.temp_order_dict)

        except KeyError:
            print("[ERROR] Order Ticket {} not found!".format(ticket))

    def push_close_order_partial(self, ticket, lots):
        """ close an order, partially"""

        try:
            self.temp_order_dict['_action'] = 'CLOSE_PARTIAL'
            self.temp_order_dict['_ticket'] = ticket
            self.temp_order_dict['_lots'] = lots

            # Execute
            self._push_command(**self.temp_order_dict)

        except KeyError:
            print("[ERROR] Order Ticket {} not found!".format(ticket))

    def push_close_magic(self, magic):
        """close orders by magic number"""

        try:
            self.temp_order_dict['_action'] = 'CLOSE_MAGIC'
            self.temp_order_dict['_magic'] = magic

            # Execute
            self._push_command(**self.temp_order_dict)

        except KeyError:
            pass

    def push_close_all(self):
        """close all orders"""
        try:
            self.temp_order_dict['_action'] = 'CLOSE_ALL'

            # Execute
            self._push_command(**self.temp_order_dict)

        except KeyError:
            pass

    def push_req_open_orders(self):
        """get all open orders"""

        try:
            self.temp_order_dict['_action'] = 'GET_OPEN_TRADES'

            # Execute
            self._push_command(**self.temp_order_dict)

        except KeyError:
            pass

    def push_req_market_data(self,
                             symbol='EURUSD',
                             timeframe=1,
                             start='2019.01.04 17:00:00',
                             end=Timestamp.now().strftime('%Y.%m.%d %H:%M:00')):

        _msg = "{};{};{};{};{}".format('DATA',
                                       symbol,
                                       timeframe,
                                       start,
                                       end)
        # Send via PUSH Socket
        self._remote_send(self.push_socket, _msg)

    def push_req_market_hist(self,
                             symbol='EURUSD',
                             timeframe=1,
                             start='2019.01.04 17:00:00',
                             end=Timestamp.now().strftime('%Y.%m.%d %H:%M:00')):
        # _end='2019.01.04 17:05:00'):

        _msg = "{};{};{};{};{}".format('HIST',
                                       symbol,
                                       timeframe,
                                       start,
                                       end)
        # Send via PUSH Socket
        self._remote_send(self.push_socket, _msg)

    def push_req_prices_track(self,
                              symbols=None):
        symbols = symbols or ['EURUSD']
        _msg = 'TRACK_PRICES'
        for s in symbols:
            _msg += ";{}".format(s)

        # Send via PUSH Socket
        self._remote_send(self.push_socket, _msg)

    def push_req_rates_track(self,
                             instruments=None):
        instruments = instruments or [('EURUSD_M1', 'EURUSD', 1)]
        _msg = 'TRACK_RATES'
        for i in instruments:
            _msg += ";{};{}".format(i[1], i[2])

        # Send via PUSH Socket
        self._remote_send(self.push_socket, _msg)

    ##########################################################################

    def subs_add_marketdata(self, symbol, string_delimiter=';'):

        # Subscribe to SYMBOL first.
        self.sub_socket.setsockopt_string(zmq.SUBSCRIBE, symbol)

        if self._DATA_THREAD is None:
            self._DATA_THREAD = Thread(target=self._receive_data, args=string_delimiter)
            self._DATA_THREAD.start()

        print("[KERNEL] Subscribed to {} MARKET updates. See self.Data_DB.".format(symbol))

    def subs_remove_marketdata(self, symbol):
        self.sub_socket.setsockopt_string(zmq.UNSUBSCRIBE, symbol)
        print('**')
        print("[KERNEL] Unsubscribing from " + symbol)
        print('**')

    def subs_remove_marketdata_all(self):
        self._set_status(False)
        self._DATA_THREAD = None

    ##########################################################################
